File: crates/atilla-tui/src/widgets/truncated_text.rs
//! Byte-exact port of `vendor/pi/packages/tui/src/components/truncated-text.ts`.

use crate::renderer::Component;
use crate::width::{truncate_to_width, visible_width};

/// Text component that truncates to fit viewport width.
pub struct TruncatedText {
    text: String,
    padding_x: usize,
    padding_y: usize,
}

impl TruncatedText {
    /// `new TruncatedText(text, paddingX = 0, paddingY = 0)`.
    pub fn new(text: &str, padding_x: usize, padding_y: usize) -> Self {
        Self {
            text: text.to_string(),
            padding_x,
            padding_y,
        }
    }
}

impl Component for TruncatedText {
    fn render(&self, width: usize) -> Vec<String> {
        let mut result: Vec<String> = Vec::new();

        // Empty line padded to width.
        let empty_line = " ".repeat(width);

        // Add vertical padding above.
        for _ in 0..self.padding_y {
            result.push(empty_line.clone());
        }

        // Calculate available width after horizontal padding.
        let available_width = width.saturating_sub(self.padding_x * 2).max(1);

        // Take only the first line (stop at newline).
        let single_line_text = match self.text.find('\n') {
            Some(idx) => &self.text[..idx],
            None => &self.text,
        };

        // Truncate text if needed (accounting for ANSI codes).
        let display_text =
            truncate_to_width(single_line_text, available_width as i64, "...", false);

        // Add horizontal padding.
        let left_padding = " ".repeat(self.padding_x);
        let right_padding = " ".repeat(self.padding_x);
        let line_with_padding = format!("{left_padding}{display_text}{right_padding}");

        // Pad line to exactly width characters.
        let line_visible_width = visible_width(&line_with_padding);
        let padding_needed = width.saturating_sub(line_visible_width);
        let final_line = format!("{line_with_padding}{}", " ".repeat(padding_needed));

        result.push(final_line);

        // Add vertical padding below.
        for _ in 0..self.padding_y {
            result.push(empty_line.clone());
        }

        result
    }

    fn invalidate(&mut self) {
        // No cached state to invalidate currently.
    }
}

/// One-shot render for a [`TruncatedText`], mirroring pi's
/// `new TruncatedText(text, paddingX, paddingY).render(width)`. Convenience
/// entry point so callers (e.g. the napi shim) need not bring the
/// [`Component`] trait into scope.
pub fn truncated_text_render(
    text: &str,
    padding_x: usize,
    padding_y: usize,
    width: usize,
) -> Vec<String> {
    TruncatedText::new(text, padding_x, padding_y).render(width)
}

File: conformance/shims/packages/tui/src/components/truncated-text.ts
// Native shim for packages/tui/src/components/truncated-text.ts, backed by the
// pidgin Rust addon (`pidgin-napi`). Installed by conformance/codegen.mjs when
// the module is marked `native` in conformance/manifest.json: the original pi
// file is preserved alongside as `truncated-text.__pi_original__.ts` and this
// shim takes its place, so pi's tests import
// `../src/components/truncated-text.ts` unchanged and hit Rust.
//
// Scope of the native flip: `TruncatedText.render` ported bit-exactly in
// `crates/pidgin-tui` (validated against pi's truncated-text.test.ts). The
// class shape — constructor `(text, paddingX = 0, paddingY = 0)`, `invalidate`,
// and `render(width)` — mirrors pi's exactly; `render` delegates to the native
// `truncatedTextRender`.

export * from "./truncated-text.__pi_original__.ts";

import type { Component } from "../tui.ts";
import { truncatedTextRender as nativeTruncatedTextRender } from "pidgin-napi";

/**
 * Text component that truncates to fit viewport width.
 */
export class TruncatedText implements Component {
	private text: string;
	private paddingX: number;
	private paddingY: number;

	constructor(text: string, paddingX: number = 0, paddingY: number = 0) {
		this.text = text;
		this.paddingX = paddingX;
		this.paddingY = paddingY;
	}

	invalidate(): void {
		// No cached state to invalidate currently
	}

	render(width: number): string[] {
		return nativeTruncatedTextRender(this.text, this.paddingX, this.paddingY, width);
	}
}
